artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ServerConsumerImpl.java

```diff
@@ -1513,7 +1524,139 @@ public String getConnectionRemoteAddress() {
       return this.session.getRemotingConnection().getTransportConnection().getRemoteAddress();
    }
 
+   @Override
+   public int getMessagesInTransitSize() {
+      return metrics.getMessagesInTransitSize();
+   }
+
+   @Override
+   public int getMessagesInTransit() {
+      return deliveringRefs.size();
+   }
+
+   @Override
+   public long getLastDeliveredTime() {
+      return metrics.getLastDeliveredTime();
+   }
+
+   @Override
+   public long getLastAcknowledgedTime() {
+      return metrics.getLastAcknowledgedTime();
+   }
+
+   @Override
+   public long getMessagesAcknowledged() {
+      return metrics.getMessagesAcknowledged();
+   }
+
+   @Override
+   public int getMessagesDeliveredSize() {
+      return metrics.getMessagesDeliveredSize();
+   }
+
+   @Override
+   public int getMessagesDelivered() {
+      return metrics.getMessagesDelivered();
+   }
+
+   @Override
+   public int getMessagesAcknowledgedAwaitingCommit() {
+      return metrics.getMessagesAcknowledgedAwaitingCommit();
+   }
+
    public SessionCallback getCallback() {
       return callback;
    }
+
+   static class ServerConsumerMetrics extends TransactionOperationAbstract {
+
+      /**
+       * Since messages can be delivered (incremented) and acknowledged (decremented) at the same time we have to protect
+       * the encode size and make it atomic. The other fields are ok since they are only accessed from a single thread.
+       */
+      private static final AtomicIntegerFieldUpdater<ServerConsumerMetrics> messagesInTransitSizeUpdater = AtomicIntegerFieldUpdater.newUpdater(ServerConsumerMetrics.class, "messagesInTransitSize");
+
+      private volatile int messagesInTransitSize = 0;
+
+      private static final AtomicIntegerFieldUpdater<ServerConsumerMetrics> messagesAcknowledgedAwaitingCommitUpdater = AtomicIntegerFieldUpdater.newUpdater(ServerConsumerMetrics.class, "messagesAcknowledgedAwaitingCommit");
+
+      private volatile int messagesAcknowledgedAwaitingCommit = 0;
+
+      private int messagesDeliveredSize = 0;
+
+      private volatile long lastDeliveredTime = 0;
+
+      private volatile long lastAcknowledgedTime = 0;
+
+      private int messagesDelivered = 0;
```

Comment:
If messagesDelivered and messagesDeliveredSize are primarily being read from other threads than the updaters, and arent synchronizing any of the same thing things used when updating the values, they should be volatile also ?

Reply:
and also FieldUpdaters?

Reply:
If it is clear that the variable updates are thread-safe (appears so in this case, 2 callers that both use 'lock') and are unlikely to become unsafe, then you dont need a FieldUpdater.